src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -413,16 +326,139 @@ public ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceCurrentSol
             public async ValueTask<Solution> GetSolutionAsync(
                 AssetProvider assetProvider,
                 Checksum solutionChecksum,
-                bool fromPrimaryBranch,
+                bool updatePrimaryBranch,
                 int workspaceVersion,
                 CancellationToken cancellationToken)
             {
                 var tuple = await _remoteWorkspace.RunWithSolutionAsync(
-                    assetProvider, solutionChecksum, workspaceVersion, fromPrimaryBranch, _ => ValueTaskFactory.FromResult(false), cancellationToken).ConfigureAwait(false);
+                    assetProvider, solutionChecksum, workspaceVersion, updatePrimaryBranch, _ => ValueTaskFactory.FromResult(false), cancellationToken).ConfigureAwait(false);
                 return tuple.solution;
             }
         }
 
+        private sealed class ChecksumToSolutionCache
+        {
+            /// <summary>
+            /// Pointer to <see cref="RemoteWorkspace._gate"/>.
+            /// </summary>
+            private readonly SemaphoreSlim _gate;
+
+            /// <summary>
+            /// The last checksum and solution requested by a service. This effectively adds an additional ref count to
+            /// one of the items in <see cref="_solutionChecksumToLazySolution"/> ensuring that the very last solution
+            /// requested is kept alive by us, even if there are no active requests currently in progress for that
+            /// solution.  That way if we have two non-concurrent requests for that same solution, with no intervening
+            /// updates, we can cache and keep the solution around instead of having to recompute it.
+            /// </summary>
+            private Checksum? _lastRequestedChecksum;
+            private ReferenceCountedDisposable<LazySolution>? _lastRequestedSolution;
+
+            /// <summary>
+            /// Mapping from solution-checksum to the solution computed for it.  This is used so that we can hold a
+            /// solution around as long as the checksum for it is being used in service of some feature operation (e.g.
+            /// classification).  As long as we're holding onto it, concurrent feature requests for the same solution
+            /// checksum can share the computation of that particular solution and avoid duplicated concurrent work.
+            /// </summary>
+            private readonly Dictionary<Checksum, ReferenceCountedDisposable<LazySolution>> _solutionChecksumToLazySolution = new();
+
+            public ChecksumToSolutionCache(SemaphoreSlim gate)
+            {
+                _gate = gate;
+            }
+
+            public async ValueTask<ReferenceCountedDisposable<LazySolution>?> TryFastGetSolutionAsync(
```

Comment:
these methods are basically the same named methods from before, just simplified now that this cache object only stores the last accessed item.